### PR TITLE
Patch SparkMagic for latest IPythonKernel compatibility

### PR DESCRIPTION
**Pull request by jaipreet-s**
**Description**
* The IPython interface was updated to return an asyncio.Future rather than a dict from version 5.1.0. This broke SparkMagic as it still expects a dictionary from the output
* This change updates the SparkMagic base kernel to expect a Future and block on its result.
* This also updates the dependencies to call out the the new IPython version dependency.
* Addresses #492 

**Testing Done**
* Unit tests added
* Validating that the kernel connects successfully
* Validating some basic Spark additional operations on an EMR cluster.

### CODE REVIEW

**Comment by jaipreet-s** [comment]
Could a maintainer have a look at this please?  The tests are breaking for for Python2 compatibiltiy, however IPython has dropped support for Python 2 since 5.0.0 [1]. Does it still make sense to support Python2 in SparkMagic?

[1] https://github.com/ipython/ipykernel/blob/master/docs/changelog.rst

**Comment by Tagar** [comment]
@jaipreet-s we still have a bunch of users who are on Python 2.7.. 
IMO we need to support Python 2 and 3 as long as PySpark itself doesn't drop Python 2 support.
It seems even upcoming Spark 3.0 release doesn't have plans to drop Python 2.7 support.

File: sparkmagic/setup.py [hunk 1]
```diff
@@ -85,7 +85,7 @@ def version(path):
           'pandas>=0.17.1',
           'numpy',
           'requests',
-          'ipykernel>=4.2.2,<5',
+          'ipykernel>=5.1.0,<6.0',
           'ipywidgets>5.0.0,<8.0',
           'notebook>=4.2,<6.0',
```
Comment (Tagar) [inline sparkmagic/setup.py]: if you're upgrading `ipykernel` from `4.2` to `5.x`, 
wouldn't you also need to upgrade `notebook` from `4.2` to `5.x` ?

Comment (arunasp) [reply sparkmagic/setup.py]: When bumping notebook version in test installation it leads to dependency hell of incompatible versions in other packages. Seems the JupyterHub docker image will also need to follow upgrades as well.

**Comment by juliusvonkohout** [comment]
What about forking into sparkmagic3 which only supports python3, drop all python2 tests and working and the myriads of unanswered pull requests and issues? @jaipreet-s I am using a sparkmagic with notebook ipywidgets plotly ipython on latest versions. and it works fine. for python2 and python3 so far. i will test whether it just breaks test or real functionality for python2.

**Comment by Tagar** [comment]
Spark still supports Python 2. 
It seems even Spark 3 release coming up in a few months has not plans of dropping Python 2 support yet.
We use both Python 3 and 2. 
Would be nice to retain support for Python 2 until end of this year at least - official EOL for Python 2.
Thank you.

**Comment by juliusvonkohout** [comment]
> Spark still supports Python 2.
> It seems even Spark 3 release coming up in a few months has no plans of dropping Python 2 support yet.
> We use both Python 3 and 2.
> Would be nice to retain support for Python 2 until end of this year at least - official EOL for Python 2.
> Thank you.

My packages still work with python2 on the cluster. but this repository badly maintained and the responses are so slow that i just forked it into https://pypi.org/project/sparkmagic3/ I merged the most important and straight forward pull requests, fixed several issues and updated all packages to the latest versions possible. We are using this now in production with livy 0.4 and 0.5.

**Comment by Tagar** [comment]
thanks @juliusvonkohout 

adding rest of `jupyter-incubator` custodians from https://github.com/jupyter-incubator page.
cc @jaipreet-s @Carreau @damianavila @jdfreder @parente @rgbkrk 

is it possible somebody would help with pushing some pull request `jupyter-incubator/sparkmagic` over to master? 

Sparkmagic has broken Python 3 support for over a year, and doesn't support several latest Apache Livy releases.. also there are a lot of PRs that don't get attention from committers for since January.

File: sparkmagic/setup.py (hunk 1 above)
Comment (rgbkrk) [reply sparkmagic/setup.py]: The kernel version doesn't have to match that of the notebook, as they're designed to be separate systems.

**Comment by rgbkrk** [comment]
cc @apetresc @aggFTW

If we don't get any responses, we'll see what we can do about current contributors becoming maintainers.

**Comment by itamarst** [comment]
I will try to review this and get it merged.

**Comment by itamarst** [comment]
I would like to get one more release out at least that supports Python 2; the current code doesn't work with Livy 0.6, so would be good to have release that does Python 2 and Livy 0.6, so people stuck with Python 2 can keep using it.

Here's the plan for Spark to drop Python 2; sounds like it's happening next year sometime:

https://spark.apache.org/news/plan-for-dropping-python-2-support.html

**Comment by itamarst** [comment]
I suspect it's possible to support both old and new versions of IPython notebook, so will see if I can manage it, based on this PR.

**Comment by itamarst** [comment]
I'm going to merge this into branch on the main repo, update it, and open new PR off that if I can get it working.

**Comment by itamarst** [comment]
So it occurs to me that Python 2 support conflates two different issues: being able support Python 2 in the notebook, and being able to send Python 2 to Spark.

It's perfectly possible for the notebook to run on Python 3, but still send Python 2 to Spark.

**Pull request merged**